### PR TITLE
WIP: MCTS, alphabeta algorithms for 2p

### PR DESCRIPTION
**Pull request by tyrcho**
@huiwang sorry for the single commit, I got a weird issue with my laptop and I had to clone again the repository.

MCTS is working well with good performance and several tests. alphabeta still requires quite some work : 

* [ ] trim branches from root
* [ ] real game tests (connect4 ? need a game with a working heuristic)
* [ ] performance checks
* [ ] iterative deepening

### CODE REVIEW

File: src/main/scala/ai/scala/fp/game/GameRules.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import scala.util.Random
+
+trait GameRules[P, S <: GameState[P], M] {
+  def initial: S
+
+  def validMoves(state: S): Seq[M]
+
+  def applyMove(state: S, move: M): S
+
+  def outcome(state: S): Outcome[P]
```
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/GameRules.scala]: I also have something similar to outcome, I called it GameResult.  I think it's a good idea to merge these concepts.

File: src/main/scala/ai/scala/fp/game/GameRules.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package ai.scala.fp.game
```
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/GameRules.scala]: having two separate packages (ai.scala.fp, com.truelaurel) is not good idea, let's find a common one. what do you think of com.codingame.scala ?

File: src/main/scala/ai/scala/fp/game/GameState.scala [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package ai.scala.fp.game
+
+trait GameState[P] {
```
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/GameState.scala]: A game state only gives the next player? why?

File: src/main/scala/ai/scala/fp/game/GameRules.scala (hunk 2 above)
Comment (tyrcho) [inline src/main/scala/ai/scala/fp/game/GameRules.scala]: hmm I agree with this one for codingame specific stuff like the bundler or arenas.
However I'd like to keep a more generic name for algorithms like genetics, MCTS, alphabeta ...

File: src/main/scala/ai/scala/fp/game/GameState.scala (hunk 3 above)
Comment (tyrcho) [inline src/main/scala/ai/scala/fp/game/GameState.scala]: Well it's the only generic part I needed so far ... all the rest is in specific implementations (see classes like [GomokuBoard](https://github.com/TrueLaurel/CodinGame-Scala-Kit/pull/26/files#diff-9f2d7ed991283dd2c62be5490fab6ffc)) . What else would you need ?

File: src/main/scala/ai/scala/fp/game/GameRules.scala (hunk 2 above)
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/GameRules.scala]: ok I see your point of view. what I've put in the com.truelaurel package are not always codingame related. let's discuss it later on how to organize our codes.

File: src/main/scala/ai/scala/fp/game/GameState.scala (hunk 3 above)
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/GameState.scala]: It was not clear my comment :) I want to understand the naming part. Given that there is one method in this trait called nextPlayer, the name GameState seems not very related. 

I have to check the example to understand better how this trait is used.

**Comment by huiwang** [comment]
@tyrcho do you want to merge this PR first and then transform the todos to other tasks that we can do in the future?

**Comment by tyrcho** [comment]
Yep, it's OK with me, better not to let branches live too long.

**Pull request merged**